runtime/src/main/java/com/dylibso/chicory/runtime/TypeValidator.java

```diff
@@ -5,128 +5,146 @@
 import com.dylibso.chicory.wasm.types.FunctionType;
 import com.dylibso.chicory.wasm.types.Instruction;
 import com.dylibso.chicory.wasm.types.OpCode;
+import com.dylibso.chicory.wasm.types.PassiveDataSegment;
 import com.dylibso.chicory.wasm.types.ValueType;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 
 // Heavily inspired by wazero
 // https://github.com/tetratelabs/wazero/blob/5a8a053bff0ae795b264de9672016745cb842070/internal/wasm/func_validation.go
+// control flow implementation follows:
+// https://webassembly.github.io/spec/core/appendix/algorithm.html
 public class TypeValidator {
 
-    private List<ValueType> valueTypeStack = new ArrayList<>();
-    private List<List<ValueType>> returns = new ArrayList<>();
-    private List<List<ValueType>> prevStack = new ArrayList<>();
+    private boolean isNum(ValueType t) {
+        return t.isNumeric() || t == ValueType.UNKNOWN;
+    }
 
-    private static <T> T peek(List<T> list) {
-        return list.get(list.size() - 1);
+    private boolean isRef(ValueType t) {
+        return t.isReference() || t == ValueType.UNKNOWN;
     }
 
-    private static <T> T pop(List<T> list) {
-        var val = list.get(list.size() - 1);
-        list.remove(list.size() - 1);
-        return val;
+    private static class CtrlFrame {
+        private final OpCode opCode;
+        private final List<ValueType> startTypes;
+        private final List<ValueType> endTypes;
+        private final int height;
+        private boolean unreachable;
+        private boolean hasElse;
```

Comment:
maybe worth adding a few comments on the meaning of these fields

Reply:
done